File: internal/core/crypto_test.go
package core

import (
	"testing"
)

func TestDeriveKey(t *testing.T) {
	salt := make([]byte, SaltSize)
	for i := range salt {
		salt[i] = byte(i)
	}

	key, err := DeriveKey("testpassword", salt, MinIterations)
	if err != nil {
		t.Fatalf("DeriveKey failed: %v", err)
	}

	if len(key) != KeySize {
		t.Errorf("Expected key length %d, got %d", KeySize, len(key))
	}

	// Same inputs should produce same key
	key2, _ := DeriveKey("testpassword", salt, MinIterations)
	for i := range key {
		if key[i] != key2[i] {
			t.Error("Same inputs produced different keys")
			break
		}
	}

	// Different password should produce different key
	key3, _ := DeriveKey("differentpassword", salt, MinIterations)
	same := true
	for i := range key {
		if key[i] != key3[i] {
			same = false
			break
		}
	}
	if same {
		t.Error("Different passwords produced same key")
	}
}

func TestDeriveKeyValidation(t *testing.T) {
	salt := make([]byte, SaltSize)

	// Too few iterations
	_, err := DeriveKey("password", salt, 1000)
	if err == nil {
		t.Error("Expected error for low iterations")
	}

	// Wrong salt size
	_, err = DeriveKey("password", []byte("short"), MinIterations)
	if err == nil {
		t.Error("Expected error for wrong salt size")
	}
}

func TestEncryptDecryptRoundTrip(t *testing.T) {
	testCases := []struct {
		name      string
		plaintext string
	}{
		{"simple", "hello world"},
		{"empty", ""},
		{"unicode", "hello world 🔐"},
		{"json", `{"key": "value", "nested": {"a": 1}}`},
		{"long", string(make([]byte, 10000))},
	}

	password := "test-master-password"
	iterations := MinIterations

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			result, err := Encrypt(tc.plaintext, password, iterations)
			if err != nil {
				t.Fatalf("Encrypt failed: %v", err)
			}

			decrypted, err := Decrypt(result, password, iterations)
			if err != nil {
				t.Fatalf("Decrypt failed: %v", err)
			}

			if decrypted != tc.plaintext {
				t.Errorf("Round trip failed: got %q, want %q", decrypted, tc.plaintext)
			}
		})
	}
}

func TestDecryptWrongPassword(t *testing.T) {
	result, _ := Encrypt("secret data", "correct-password", MinIterations)

	_, err := Decrypt(result, "wrong-password", MinIterations)
	if err == nil {
		t.Error("Expected error when decrypting with wrong password")
	}
}

func TestEncryptProducesUniqueOutput(t *testing.T) {
	plaintext := "same plaintext"
	password := "same password"

	result1, _ := Encrypt(plaintext, password, MinIterations)
	result2, _ := Encrypt(plaintext, password, MinIterations)

	// Salt should differ
	if result1.Salt == result2.Salt {
		t.Error("Expected different salts")
	}

	// IV should differ
	if result1.IV == result2.IV {
		t.Error("Expected different IVs")
	}

	// Ciphertext should differ
	if result1.Ciphertext == result2.Ciphertext {
		t.Error("Expected different ciphertexts")
	}

	// But both should decrypt to same plaintext
	d1, _ := Decrypt(result1, password, MinIterations)
	d2, _ := Decrypt(result2, password, MinIterations)
	if d1 != d2 || d1 != plaintext {
		t.Error("Both should decrypt to original plaintext")
	}
}

File: legacy/src/crypto.test.ts
/**
 * Unit tests for crypto module
 */

import { test } from 'node:test';
import * as assert from 'node:assert/strict';
import { encrypt, decrypt, deriveKey, verifyPassword, EncryptionResult } from './crypto.js';

test('Crypto Module', async (t) => {
  await t.test('deriveKey generates consistent keys for same password and salt', () => {
    const password = 'test-password';
    const { key: key1, salt } = deriveKey(password);
    const { key: key2 } = deriveKey(password, salt, 100000);

    assert.deepEqual(key1, key2, 'Keys should match when using same salt and iterations');
  });

  await t.test('deriveKey generates different keys for different passwords', () => {
    const { key: key1, salt } = deriveKey('password1');
    const { key: key2 } = deriveKey('password2', salt, 100000);

    assert.notDeepEqual(key1, key2, 'Keys should differ for different passwords');
  });

  await t.test('deriveKey generates different salt on each call', () => {
    const { salt: salt1 } = deriveKey('password');
    const { salt: salt2 } = deriveKey('password');

    assert.notDeepEqual(salt1, salt2, 'Salts should be different');
  });

  await t.test('deriveKey requires minimum iterations for security', () => {
    assert.throws(
      () => deriveKey('password', undefined, 50000),
      /PBKDF2 iterations must be at least 100,000/,
      'Should reject iterations < 100,000'
    );
  });

  await t.test('encrypt creates valid encryption result', () => {
    const result = encrypt('secret data', 'password');

    assert.ok(result.ciphertext, 'Should have ciphertext');
    assert.ok(result.authTag, 'Should have authTag');
    assert.ok(result.iv, 'Should have iv');
    assert.ok(result.salt, 'Should have salt');

    // Check that they're base64 encoded
    assert.ok(Buffer.from(result.ciphertext, 'base64').length > 0);
    assert.ok(Buffer.from(result.authTag, 'base64').length > 0);
    assert.ok(Buffer.from(result.iv, 'base64').length > 0);
    assert.ok(Buffer.from(result.salt, 'base64').length > 0);
  });

  await t.test('encrypt produces different ciphertext for same plaintext', () => {
    const result1 = encrypt('secret data', 'password');
    const result2 = encrypt('secret data', 'password');

    assert.notEqual(result1.ciphertext, result2.ciphertext, 'Should produce different ciphertexts (due to random IV and salt)');
  });

  await t.test('decrypt recovers original plaintext', () => {
    const plaintext = 'secret data 123!@#';
    const password = 'my-password';

    const encrypted = encrypt(plaintext, password);
    const decrypted = decrypt(encrypted, password);

    assert.equal(decrypted, plaintext, 'Decrypted text should match original');
  });

  await t.test('decrypt fails with wrong password', () => {
    const encrypted = encrypt('secret data', 'correct-password');

    assert.throws(
      () => decrypt(encrypted, 'wrong-password'),
      /Decryption failed/,
      'Should fail with wrong password'
    );
  });

  await t.test('decrypt handles empty string', () => {
    const plaintext = '';
    const encrypted = encrypt(plaintext, 'password');
    const decrypted = decrypt(encrypted, 'password');

    assert.equal(decrypted, plaintext, 'Should handle empty strings');
  });

  await t.test('decrypt handles large data', () => {
    const largeData = 'x'.repeat(100000);
    const encrypted = encrypt(largeData, 'password');
    const decrypted = decrypt(encrypted, 'password');

    assert.equal(decrypted, largeData, 'Should handle large data');
  });

  await t.test('decrypt handles special characters', () => {
    const plaintext = '{"key": "value", "emoji": "🔒", "unicode": "你好"}';
    const encrypted = encrypt(plaintext, 'password');
    const decrypted = decrypt(encrypted, 'password');

    assert.equal(decrypted, plaintext, 'Should preserve special characters');
  });

  await t.test('verifyPassword returns true for correct password', () => {
    const encrypted = encrypt('data', 'correct-password');
    const isCorrect = verifyPassword(encrypted, 'correct-password');

    assert.equal(isCorrect, true, 'Should return true for correct password');
  });

  await t.test('verifyPassword returns false for incorrect password', () => {
    const encrypted = encrypt('data', 'correct-password');
    const isCorrect = verifyPassword(encrypted, 'wrong-password');

    assert.equal(isCorrect, false, 'Should return false for wrong password');
  });

  await t.test('encrypt with custom iterations', () => {
    const plaintext = 'secret';
    const encrypted = encrypt(plaintext, 'password', 150000);
    const decrypted = decrypt(encrypted, 'password', 150000);

    assert.equal(decrypted, plaintext, 'Should work with custom iterations');
  });

  await t.test('decrypt fails with corrupted ciphertext', () => {
    const encrypted = encrypt('secret', 'password');
    const corrupted: EncryptionResult = {
      ...encrypted,
      ciphertext: Buffer.from('corrupted-data').toString('base64'),
    };

    assert.throws(
      () => decrypt(corrupted, 'password'),
      /Decryption failed/,
      'Should fail with corrupted ciphertext'
    );
  });

  await t.test('decrypt fails with corrupted auth tag', () => {
    const encrypted = encrypt('secret', 'password');
    const corrupted: EncryptionResult = {
      ...encrypted,
      authTag: Buffer.from('bad-tag').toString('base64'),
    };

    assert.throws(
      () => decrypt(corrupted, 'password'),
      /Decryption failed/,
      'Should fail with corrupted auth tag'
    );
  });

  await t.test('encrypt and decrypt preserve JSON structure', () => {
    const jsonData = {
      name: 'test',
      secrets: {
        api_key: 'sk-123',
        token: 'tk-456',
      },
      count: 42,
    };
    const plaintext = JSON.stringify(jsonData);

    const encrypted = encrypt(plaintext, 'password');
    const decrypted = decrypt(encrypted, 'password');
    const parsed = JSON.parse(decrypted);

    assert.deepEqual(parsed, jsonData, 'Should preserve JSON structure');
  });
});
